plasma_framework/contracts/src/transactions/eip712Libs/PaymentEip712Lib.sol

```diff
@@ -0,0 +1,120 @@
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+import "../PaymentTransactionModel.sol";
+import "../outputs/PaymentOutputModel.sol";
+import "../../utils/UtxoPosLib.sol";
+
+library PaymentEip712Lib {
+    using UtxoPosLib for UtxoPosLib.UtxoPos;
+
+    uint8 constant public MAX_INPUT_NUM = 4;
+    uint8 constant public MAX_OUTPUT_NUM = 4;
+
+    struct Constants {
+        bytes2 EIP191_PREFIX;
+        bytes32 EIP712_DOMAIN_HASH;
+        bytes32 TX_TYPE_HASH;
+        bytes32 INPUT_TYPE_HASH;
+        bytes32 OUTPUT_TYPE_HASH;
+        bytes32 DOMAIN_SEPARATOR;
+    }
+
+    function initConstants(address _verifyingContract) internal pure returns (Constants memory) {
+        bytes2 EIP191_PREFIX = "\x19\x01";
+        bytes32 EIP712_DOMAIN_HASH = keccak256(abi.encodePacked(
+            "EIP712Domain(string name,string version,address verifyingContract,bytes32 salt)"
+        ));
+        bytes32 TX_TYPE_HASH = keccak256(abi.encodePacked(
```

Comment:
yeah, I was pretty confused on what should we do for deposit when I was writing this.
So you're proposing 4 inputs with EMPTY value and 3 more outputs with empty value?

Personally I find our current deposit input confusing as it explicits one empty value instead of pure empty input. I think we can go either make it no input or make it 4 empty. As for output probably it does not matters much as there would be payment transaction without using all 4 outputs anyway. Unless we want to enforce all transaction need to fill in input/output with empty value?

Reply:
In current design we enforce transaction to conform to the EIP format presented here, that is each transaction has 4 inputs and 4 outputs (some of them may be fillers - all 0s). 

I think the good way for deposits it to make it no input, but if we make it no input then they do not conform to EIP format. That might be confusing for users. Solution would be to have a different EIP format for deposits or use dynamic arrays in EIP (I don't know how to do it, I was trying to come up with something for my IFE, but gave up and saved it for later).

Reply:
AFAIK, metamask would be a blocker to sign dynamic array (though not really EIP712 protocol issue itself): https://github.com/omisego/elixir-omg/issues/551#issuecomment-479461060

The simplest solution now is we never put empty/dummy value in transaction and keep it 4 inputs 4 outputs, and only do that right before singing (transform to EIP712 transaction format). For me, as long as it is 1:1 mapping between two tx model I think it should be fine.

I do agree it might looks weird when user needs to sign a bunch of 0s (lol). Curious have we got any feedback on this @Pongch from user using our current EIP712 solution?